README.md
# TechDivision_RewriteModule

[![Latest Stable Version](https://poser.pugx.org/techdivision/rewritemodule/v/stable.png)](https://packagist.org/packages/techdivision/rewritemodule) [![Total Downloads](https://poser.pugx.org/techdivision/rewritemodule/downloads.png)](https://packagist.org/packages/techdivision/rewritemodule) [![Dependencies Status](https://depending.in/techdivision/TechDivision_RewriteModule.png)](http://depending.in/techdivision/TechDivision_RewriteModule) [![License](https://poser.pugx.org/techdivision/rewritemodule/license.png)](https://packagist.org/packages/techdivision/rewritemodule) [![Build Status](https://travis-ci.org/techdivision/TechDivision_RewriteModule.png)](https://travis-ci.org/techdivision/TechDivision_RewriteModule)[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/techdivision/TechDivision_RewriteModule/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/techdivision/TechDivision_RewriteModule/?branch=master)[![Code Coverage](https://scrutinizer-ci.com/g/techdivision/TechDivision_RewriteModule/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/techdivision/TechDivision_RewriteModule/?branch=master)

## Introduction

A simple rewrite module for PHP based web servers which uses a self made structure for usable rules. It can be 
used similar to [`Apaches mod_rewrite`](http://httpd.apache.org/docs/current/mod/mod_rewrite.html) and provides
rewriting and redirecting capabilities.

## Installation

If you want to use the rewrite module with your application you can install it by adding

```sh
{
    "require": {
        "techdivision/rewritemodule": "dev-master"
    },
}
```

to your ```composer.json``` and invoke ```composer update``` in your project.

## Documentation
Our documentation moved!
Have a look [here](<http://docs.appserver.io/master/components/servers/webserver/modules/rewritemodule.html>)


CHANGELOG.md
# Version 0.6.4

## Bugfixes

* Removed SSL context backreferences as they influenced the request context backreferences

## Features

* None

# Version 0.6.3

## Bugfixes

* None

## Features

* Add possibility to set custom status codes (between 300 and 399) for redirects => analog apache mod_rewrite
* Add support for redirects relative to the current host => analog apache mod_rewrite

# Version 0.6.2

## Bugfixes

* None

## Features

* Refactoring ANT PHPUnit execution process
* Composer integration by optimizing folder structure (move bootstrap.php + phpunit.xml.dist => phpunit.xml)
* Switch to new appserver-io/build build- and deployment environment

tests/bootstrap.php
<?php
/**
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * PHP version 5
 *
 * @category  WebServer
 * @package   TechDivision_RewriteModule
 * @author    Bernhard Wick <b.wick@techdivision.com>
 * @copyright 2014 TechDivision GmbH - <info@techdivision.com>
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link      http://www.techdivision.com/
 */
$loader = require '${php-target.dir}/vendor/autoload.php';
$loader->add('TechDivision\\RewriteModule\\', array(
        '${php-target.dir}/vendor/techdivision/rewritemodule/src',
        '${php-target.dir}/vendor/techdivision/rewritemodule/tests'
    ));
$loader->add('TechDivision\\Http\\', '${php-target.dir}/vendor/techdivision/http/src');
$loader->add('TechDivision\\Server\\', '${php-target.dir}/vendor/techdivision/server/src');
$loader->add('TechDivision\\WebServer\\', '${php-target.dir}/vendor/techdivision/webserver/src');

// We have to create a symbolic link for our tests here as we cannot copy them via ant
symlink(
    '${php-target.dir}/vendor/techdivision/rewritemodule/src/TechDivision/RewriteModule/_files/html/index.html',
    '${php-target.dir}/vendor/techdivision/rewritemodule/src/TechDivision/RewriteModule/_files/html/symlink.html'
);


tests/TechDivision/RewriteModule/Mock/MockRule.php
<?php
/**
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * PHP version 5
 *
 * @category   WebServer
 * @package    TechDivision_RewriteModule
 * @subpackage Mock
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */

namespace TechDivision\RewriteModule\Mock;

use TechDivision\RewriteModule\Entities\Rule;

/**
 * TechDivision\RewriteModule\Mock\MockRule
 *
 * Mocks the Rule class to expose additional and hidden functionality
 *
 * @category   WebServer
 * @package    TechDivision_RewriteModule
 * @subpackage Mock
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */
class MockRule extends Rule
{
    /**
     * Used to open up the parent's sortFlags() method for testing
     *
     * @param string $flagString The unsorted string of flags
     *
     * @return array
     */
    public function sortFlags($flagString)
    {
        return parent::sortFlags($flagString);
    }

    /**
     * Getter function for the protected $type member
     *
     * @return string
     */
    public function getType()
    {
        return $this->type;
    }

    /**
     * Getter function for the protected $sortedFlags member
     *
     * @return array
     */
    public function getSortedFlags()
    {
        return $this->sortedFlags;
    }
}


tests/TechDivision/RewriteModule/_files/flag_data.php
<?php
/**
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * PHP version 5
 *
 * @category  WebServer
 * @package   TechDivision_RewriteModule
 * @author    Bernhard Wick <b.wick@techdivision.com>
 * @copyright 2014 TechDivision GmbH - <info@techdivision.com>
 * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link      http://www.techdivision.com/
 */

/**
 * This list contains rewrite rules as they would be used within our server configuration.
 * To make tests independent from config parsing we have to provide them already split up.
 *
 * All entries follow the structure below:
 *
 *  '<DATASET_NAME>' => array( // a specific name of the data set
 *      'redirect' => true, // optional, whether or not the listed rewrites are redirects (have to be tested differently)
 *      'redirectAs' => 301, // optional, might contain a custom status code used for redirects (between 300 and 399)
 *      'rules' => array( // array of rule arrays
 *          array(
 *              'condition' => '.*', // condition as one would use it within a rewrite definition
 *              'target' => 'https://www.google.com', // target as one would use it within a rewrite definition
 *              'flag' => 'R' // flag string as one would use it within a rewrite definition
 *          )
 *      ),
 *      'map' => array( // this map contains URI/URL pairs of the sort "incoming URI" => "expected URI/URL after rewrite"
 *          '/html/index.html' => 'https://www.google.com'
 *      )
 *  ),
 *
 * @var array $ruleSets The rewrite rule sets this test is based on
 */
$ruleSets = array(
    'LFlag' => array(
        'rules' => array(
            array(
                'condition' => '.*',
                'target' => '',
                'flag' => 'L'
            ),
            array(
                'condition' => '(.*)',
                'target' => '/ERROR',
                'flag' => ''
            )
        ),
        'map' => array(
            '/testUri' => '/testUri'
        )
    ),
    'RFlag' => array(
        'redirect' => true,
        'rules' => array(
            array(
                'condition' => '.*',
                'target' => 'https://www.google.com',
                'flag' => 'R'
            )
        ),
        'map' => array(
            '/testUri' => 'https://www.google.com'
        )
    ),
    'mixedFlags' => array(
        'redirect' => true,
        'rules' => array(
            array(
                'condition' => '.*',
                'target' => 'https://www.google.com',
                'flag' => 'R,L'
            ),
            array(
                'condition' => '.*',
                'target' => '/ERROR',
                'flag' => ''
            )
        ),
        'map' => array(
            '/testUri' => 'https://www.google.com'
        )
    )
);


src/TechDivision/RewriteModule/Dictionaries/RuleFlags.php
<?php
/**
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * PHP version 5
 *
 * @category   Appserver
 * @package    TechDivision_RewriteModule
 * @subpackage Dictionaries
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php
 *             Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */

namespace TechDivision\RewriteModule\Dictionaries;

/**
 * TechDivision\RewriteModule\Dictionaries\RuleFlags
 *
 * <TODO CLASS DESCRIPTION>
 *
 * @category   Appserver
 * @package    TechDivision_RewriteModule
 * @subpackage Dictionaries
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php
 *             Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */
class RuleFlags
{
    /**
     * Defines constant for flags we might use within the rule's flag field
     *
     * @var string
     */
    const REDIRECT = 'R';
    const MAP = 'M';
    const LAST = 'L';
}


src/TechDivision/RewriteModule/Dictionaries/ConditionActions.php
<?php
/**
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * PHP version 5
 *
 * @category   Appserver
 * @package    TechDivision_RewriteModule
 * @subpackage Dictionaries
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */

namespace TechDivision\RewriteModule\Dictionaries;

/**
 * TechDivision\RewriteModule\Dictionaries\ConditionActions
 *
 * A dictionary for actions a condition might use for testing if it matches.
 * Basic stuff used parallel to htaccess features
 *
 * @category   Appserver
 * @package    TechDivision_RewriteModule
 * @subpackage Dictionaries
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */
class ConditionActions
{
    /**
     * Lexical string comparison where $operand is less than the given second string
     *
     * @var string
     */
    const STR_LESS= '<';

    /**
     * Lexical string comparison where $operand is greater than the given second string
     *
     * @var string
     */
    const STR_GREATER = '>';

    /**
     * Lexical string comparison where $operand equals the given second string
     *
     * @var string
     */
    const STR_EQUAL = '=';

    /**
     * Tests if $operand is a valid directory
     *
     * @var string
     */
    const IS_DIR= '-d';

    /**
     * Tests if $operand is a real file
     *
     * @var string
     */
    const IS_FILE = '-f';

    /**
     * Tests if $operand is a real file with a size bigger than 0
     *
     * @var string
     */
    const IS_USED_FILE = '-s';

    /**
     * Tests if $operand is a symbolic link
     *
     * @var string
     */
    const IS_LINK = '-l';

    /**
     * Tests if $operand is a file system structure which is flagged as executable
     *
     * @var string
     */
    const IS_EXECUTABLE = '-x';

    /**
     * Tests for any PCRE regex
     *
     * @var string
     */
    const REGEX = '-r';
}


src/TechDivision/RewriteModule/Entities/Rule.php
<?php
/**
 * NOTICE OF LICENSE
 *
 * This source file is subject to the Open Software License (OSL 3.0)
 * that is available through the world-wide-web at this URL:
 * http://opensource.org/licenses/osl-3.0.php
 *
 * PHP version 5
 *
 * @category   Webserver
 * @package    TechDivision_RewriteModule
 * @subpackage Entities
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */

namespace TechDivision\RewriteModule\Entities;

use TechDivision\Http\HttpProtocol;
use TechDivision\Http\HttpResponseInterface;
use TechDivision\Http\HttpResponseStates;
use TechDivision\RewriteModule\Dictionaries\RuleFlags;
use TechDivision\Server\Dictionaries\ServerVars;
use TechDivision\Server\Interfaces\RequestContextInterface;

/**
 * TechDivision\RewriteModule\Entities\Rule
 *
 * This class provides an object based representation of a rewrite rule, including logic for testing, applying
 * and handeling conditions.
 *
 * @category   Webserver
 * @package    TechDivision_RewriteModule
 * @subpackage Entities
 * @author     Bernhard Wick <b.wick@techdivision.com>
 * @copyright  2014 TechDivision GmbH - <info@techdivision.com>
 * @license    http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
 * @link       http://www.techdivision.com/
 */
class Rule
{
    /**
     * The allowed values the $type member my assume
     *
     * @var array $allowedTypes
     */
    protected $allowedTypes = array();

    /**
     * The type of rule we have. This might be "relative", "absolute" or "url"
     *
     * @var string $type
     */
    protected $type;

    /**
     * The condition string
     *
     * @var string $conditionString
     */
    protected $conditionString;

    /**
     * The sorted conditions we have to check
     *
     * @var array $sortedConditions
     */
    protected $sortedConditions = array();

    /**
     * Will hold the backreferences of the condition(s) which matched
     *
     * @var array $matchingBackreferences
     */
    protected $matchingBackreferences = array();

    /**
     * The target to rewrite the REDIRECT_URL to
     *
     * @var string $target
     */
    protected $target;

    /**
     * The flag we have to take into consideration when working with the rule
     *
     * @var string $flagString
     */
    protected $flagString;

    /**
     * All flags sorted and brought in relation with their potential parameters
     *
     * @var array $sortedFlags
     */
    protected $sortedFlags;

    /**
     * The default operand we will check all conditions against if none was given explicitly
     *
     * @const string DEFAULT_OPERAND
     */
    const DEFAULT_OPERAND = '@$X_REQUEST_URI';

    /**
     * This constant by which conditions are separated and marked as or-combined
     *
     * @const string CONDITION_OR_DELIMITER
     */
    const CONDITION_OR_DELIMITER = '{OR}';

    /**
     * This constant by which conditions are separated and marked as and-combined (the default)
     *
     * @const string CONDITION_AND_DELIMITER
     */
    const CONDITION_AND_DELIMITER = '{AND}';

    /**
     * This constant is used to separate flags from each other
     *
     * @const string FLAG_DELIMITER
     */
    const FLAG_DELIMITER = ',';

    /**
     * This constant is used to separate flags from their potential parameters
     *
     * @const string FLAG_PARAMETER_ASSIGNER
     */
    const FLAG_PARAMETER_ASSIGNER = '=';

    /**
     * Default constructor
     *
     * @param string       $conditionString Condition string e.g. "^_Resources/.*" or "-f{OR}-d{OR}-d@$REQUEST_FILENAME"
     * @param string|array $target          The target to rewrite to, might be null if we should do nothing
     * @param string       $flagString      A flag string which might be added to to the rule e.g. "L" or "C,R"
     */
    public function __construct($conditionString, $target, $flagString)
    {
        // Set the raw string properties and append our default operand to the condition string
        $this->conditionString = $conditionString;
        $conditionString .= $this->getDefaultOperand();
        $this->target = $target;
        $this->flagString = $flagString;

        // Get the sorted flags, should be easy to break up
        $this->sortedFlags = $this->sortFlags($this->flagString);

        // Set our default values here
        $this->allowedTypes = array('relative', 'absolute', 'url');
        $this->matchingBackreferences = array();

        // filter the condition string using our regex, but first of all we will append the default operand
        $conditionPieces = array();
        preg_match_all('`(.*?)@(\$[0-9a-zA-Z_]+)`', $conditionString, $conditionPieces);
        // The first index is always useless, unset it to avoid confusion
        unset($conditionPieces[0]);

        // Conditions are kind of sorted now, we can split them up into condition actions and their operands
        $conditionActions = $conditionPieces[1];
        $conditionOperands = $conditionPieces[2];

        // Iterate over the condition piece arrays, trim them and build our array of sorted condition objects
        for ($i = 0; $i < count($conditionActions); $i++) {

            // Trim whatever we got here as the string might be a bit dirty
            $actionString = trim(
                $conditionActions[$i],
                self::CONDITION_OR_DELIMITER . '|' . self::CONDITION_AND_DELIMITER
            );

            // Collect all and-combined pieces of the conditionstring
            $andActionStringPieces = explode(self::CONDITION_AND_DELIMITER, $actionString);

            // Iterate through them and build up conditions or or-combined condition groups
            foreach ($andActionStringPieces as $andActionStringPiece) {

                // Everything is and-combined (plain array) unless combined otherwise (with a "{OR}" symbol)
                // If we find an or-combination we will make a deeper array within our sorted condition array
                if (strpos($andActionStringPiece, self::CONDITION_OR_DELIMITER) !== false) {

                    // Collect all or-combined conditions into a separate array
                    $actionStringPieces = explode(self::CONDITION_OR_DELIMITER, $andActionStringPiece);

                    // Iterate over the pieces we found and create a condition for each of them
                    $entry = array();
                    foreach ($actionStringPieces as $actionStringPiece) {

                        // Get a new condition instance
                        $entry[] = new Condition($conditionOperands[$i], $actionStringPiece);
                    }

                } else {

                    // Get a new condition instance
                    $entry = new Condition($conditionOperands[$i], $andActionStringPiece);
                }

                $this->sortedConditions[] = $entry;
            }
        }
    }

    /**
     * Sort the flag string as we will have a bad time parsing the string over and over again
     *
     * @param string $flagString The unsorted string of flags
     *
     * @return array
     */
    protected function sortFlags($flagString)
    {
        $flags = array();
        foreach (explode(self::FLAG_DELIMITER, $flagString) as $flag) {

            $flagPieces = explode(self::FLAG_PARAMETER_ASSIGNER, $flag);

            // Set the pieces (if any)
            if (array_key_exists(1, $flagPieces)) {

                $flags[$flagPieces[0]] = $flagPieces[1];

            } else {

                $flags[$flagPieces[0]] = null;
            }
        }

        return $flags;
    }

    /**
     * Will return the default operand of this action
     *
     * @return string
     */
    protected function getDefaultOperand()
    {
        return self::DEFAULT_OPERAND;
    }

    /**
     * Will resolve the directive's parts by substituting placeholders with the corresponding backreferences
     *
     * @param array $backreferences The backreferences used for resolving placeholders
     *
     * @return void
     */
    public function resolve(array $backreferences)
    {
        // We have to resolve backreferences in three steps.
        // First of all we have to resolve the backreferences based on the server vars
        $this->resolveConditions($backreferences);

        // Second we have to produce the regex based backreferences from the now semi-resolved conditions
        $conditionBackreferences = $this->getBackreferences();

        // Last but not least we have to resolve the conditions against the regex backreferences
        $this->resolveConditions($conditionBackreferences);
    }

    /**
     * Will resolve the directive's parts by substituting placeholders with the corresponding backreferences
     *
     * @param array $backreferences The backreferences used for resolving placeholders
     *
     * @return void
     */
    protected function resolveConditions(array $backreferences)
    {
        // Iterate over all conditions and resolve them too
        foreach ($this->sortedConditions as $key => $sortedCondition) {

            // If we got an array we have to iterate over it separately, but be aware they are or-combined
            if (is_array($sortedCondition)) {

                // These are or-combined conditions but we have to resolve them too
                foreach ($sortedCondition as $orKey => $orCombinedCondition) {

                    // Resolve the condition
                    $orCombinedCondition->resolve($backreferences);
                    $this->sortedConditions[$key][$orKey] = $orCombinedCondition;
                }

            } else {

                // Resolve the condition
                $sortedCondition->resolve($backreferences);
                $this->sortedConditions[$key] = $sortedCondition;
            }
        }
    }

    /**
     * Will return true if the rule applies, false if not
     *
     * @return bool
     */
    public function matches()
    {
        // We will iterate over all conditions (and the or-combined condition groups) and if there is a non-matching
        // condition or condition group we will fail
        foreach ($this->sortedConditions as $sortedCondition) {

            // If we got an array we have to iterate over it separately, but be aware they are or-combined
            if (is_array($sortedCondition)) {

                // These are or-combined conditions, so break if we match one
                $orGroupMatched = false;
                foreach ($sortedCondition as $orCombinedCondition) {

                    if ($orCombinedCondition->matches()) {

                        $orGroupMatched = true;
                        $this->matchingBackreferences = array_merge(
                            $this->matchingBackreferences,
                            $orCombinedCondition->getBackreferences()
                        );
                        break;
                    }
                }

                // Did one condition within this group match?
                if ($orGroupMatched === false) {

                    return false;
                }
            } elseif (!$sortedCondition->matches()) {
                // The single conditions have to match as they are and-combined
                return false;
            } else {

                $this->matchingBackreferences = array_merge(
                    $this->matchingBackreferences,
                    $sortedCondition->getBackreferences()
                );
            }
        }

        // We are still here, this sounds good
        return true;
    }

    /**
     * Initiates the module
     *
     * @param \TechDivision\Server\Interfaces\RequestContextInterface $requestContext       The request's context
     * @param \TechDivision\Http\HttpResponseInterface                $response             The response instance
     * @param array                                                   $serverBackreferences Server backreferences
     *
     * @throws \InvalidArgumentException
     *
     * @return boolean
     */
    public function apply(
        RequestContextInterface $requestContext,
        HttpResponseInterface $response,
        array $serverBackreferences
    ) {

        // First of all we have to resolve the target string with the backreferences of the matching condition
        // Separate the keys from the values so we can use them in str_replace
        // And also mix in the server's backreferences for good measure
        $this->matchingBackreferences = array_merge($this->matchingBackreferences, $serverBackreferences);
        $backreferenceHolders = array_keys($this->matchingBackreferences);
        $backreferenceValues = array_values($this->matchingBackreferences);

        // If we got a target map (flag "M") we have to resolve the target string we have to use first
        // The following checks will be treated as an additional condition
        if (array_key_exists(RuleFlags::MAP, $this->sortedFlags) && !empty($this->sortedFlags[RuleFlags::MAP])) {

            // Get our map key for better readability
            $mapKey = str_replace(
                array_keys($this->matchingBackreferences),
                $this->matchingBackreferences,
                $this->sortedFlags[RuleFlags::MAP]
            );

            // Still here? That sounds good. Get the needed target string now
            if (isset($this->target[$mapKey])) {

                $this->target = $this->target[$mapKey];

            } else {

                // Empty string, we will do nothing with this rule
                $this->target = '';

                // Also clear any L-flag we might find, we could not find what we are looking for so we should not end
                if (array_key_exists(RuleFlags::LAST, $this->sortedFlags)) {

                    unset($this->sortedFlags[RuleFlags::LAST]);
                }
            }
        }

        // Back to our rule...
        // If the target string is empty we do not have to do anything
        if (!empty($this->target)) {

            // Just make sure that you check for the existence of the query string first, as it might not be set
            $queryFreeRequestUri = $requestContext->getServerVar(ServerVars::X_REQUEST_URI);
            if ($requestContext->hasServerVar(ServerVars::QUERY_STRING)) {

                $queryFreeRequestUri = str_replace(
                    '?' . $requestContext->getServerVar(ServerVars::QUERY_STRING),
                    '',
                    $queryFreeRequestUri
                );

                // Set the "redirect" query string as a backup as we might change the original
                $requestContext->setServerVar(
                    'REDIRECT_QUERY_STRING',
                    $requestContext->getServerVar(ServerVars::QUERY_STRING)
                );
            }
            $requestContext->setServerVar('REDIRECT_URL', $queryFreeRequestUri);

            // Substitute the backreferences in our operation
            $this->target = str_replace($backreferenceHolders, $backreferenceValues, $this->target);

            // We have to find out what type of rule we got here
            if (is_readable($this->target)) {

                // We have an absolute file path!
                $this->type = 'absolute';

                // Set the REQUEST_FILENAME path
                $requestContext->setServerVar(ServerVars::REQUEST_FILENAME, $this->target);

            } elseif (filter_var($this->target, FILTER_VALIDATE_URL) !== false) {

                // We have a complete URL!
                $this->type = 'url';

            } else {
                // Last but not least we might have gotten a relative path (most likely)
                // Build up the REQUEST_FILENAME from DOCUMENT_ROOT and X_REQUEST_URI (without the query string)
                $this->type = 'relative';

                $requestContext->setServerVar(
                    ServerVars::SCRIPT_FILENAME,
                    $requestContext->getServerVar(ServerVars::DOCUMENT_ROOT) . DIRECTORY_SEPARATOR . $this->target
                );
                $requestContext->setServerVar(ServerVars::SCRIPT_NAME, $this->target);

                // Setting the X_REQUEST_URI for internal communication
                // Requested uri always has to begin with a slash
                $this->target = '/' . ltrim($this->target, '/');
                $requestContext->setServerVar(ServerVars::X_REQUEST_URI, $this->target);

                // Only change the query string if we have one in our target string
                if (strpos($this->target, '?') !== false) {

                    $requestContext->setServerVar(ServerVars::QUERY_STRING, substr(strstr($this->target, '?'), 1));
                }
            }

            // do we have to make a redirect?
            // if so we have to have to set the status code accordingly and dispatch the response
            if (array_key_exists(RuleFlags::REDIRECT, $this->sortedFlags)) {

                $this->prepareRedirect($requestContext, $response);
            }

            // Lets tell them that we successfully made a redirect
            $requestContext->setServerVar(ServerVars::REDIRECT_STATUS, '200');
        }
        // If we got the "LAST"-flag we have to end here, so return false
        if (array_key_exists(RuleFlags::LAST, $this->sortedFlags)) {

            return false;
        }

        // Still here? That sounds good
        return true;
    }

    /**
     * Will prepare a response for a redirect.
     * This includes setting the new target, the appropriate status code and dispatching it to break the
     * module chain
     *
     * @param \TechDivision\Server\Interfaces\RequestContextInterface $requestContext The request's context
     * @param \TechDivision\Http\HttpResponseInterface                $response       The response instance to be prepared
     *
     * @return void
     */
    protected function prepareRedirect($requestContext, $response)
    {
        // if we got a specific status code we have to filter it and apply it if possible
        $statusCode = 301;
        $proposedStatusCode = $this->sortedFlags[RuleFlags::REDIRECT];
        if (is_numeric($proposedStatusCode) && $proposedStatusCode >= 300 && $proposedStatusCode < 400) {

            $statusCode = $proposedStatusCode;
        }

        // there might be work to be done depending on whether or not we got a complete URL
        if ($this->type === 'relative') {

            $newTarget = $requestContext->getServerVar(ServerVars::REQUEST_SCHEME);
            $newTarget .= '://';
            $newTarget .= $requestContext->getServerVar(ServerVars::HTTP_HOST);
            $this->target = $newTarget . $this->getTarget();
        }

        // set enhance uri to response
        $response->addHeader(HttpProtocol::HEADER_LOCATION, $this->target);
        // send redirect status
        $response->setStatusCode($statusCode);
        // set response state to be dispatched after this without calling other modules process
        $response->setState(HttpResponseStates::DISPATCH);
    }

    /**
     * Will collect all backreferences based on regex typed conditions
     *
     * @return array
     */
    public function getBackreferences()
    {
        // Iterate over all conditions and collect their backreferences
        $backreferences = array();
        foreach ($this->sortedConditions as $key => $sortedCondition) {

            // If we got an array we have to iterate over it separately, but be aware they are or-combined
            if (is_array($sortedCondition)) {

                // These are or-combined conditions but we have to resolve them too
                foreach ($sortedCondition as $orCombinedCondition) {

                    // Get the backreferences of this condition
                    $backreferences = array_merge(
                        $backreferences,
                        $orCombinedCondition->getBackreferences()
                    );
                }

            } else {

                // Get the backreferences of this condition
                $backreferences = array_merge(
                    $backreferences,
                    $sortedCondition->getBackreferences()
                );
            }
        }

        return $backreferences;
    }

    /**
     * Getter function for the condition string
     *
     * @return string
     */
    public function getConditionString()
    {
        return $this->conditionString;
    }

    /**
     * Getter function for the flag string
     *
     * @return string
     */
    public function getFlagString()
    {
        return $this->flagString;
    }

    /**
     * Getter function for the target string
     *
     * @return string
     */
    public function getTarget()
    {
        return $this->target;
    }
}
